Xenoborgs part 7

Samuka-C opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
This is it, the final xenoborg part

A few things from last PRs got changed, such as:
- mothership core now has modules with hands (one for materials, one for brains, one for modules)
- removed the hands from the mothership core sprite
- space carps will now attack xenoborgs
- xenoborgs drop a pinpointer to the mothership core
- the space movement module now includes a pinpointer

What this PR adds
- new Xenoborg rule, which is a subgamemode, roundstart, similar to the sub-wizard.
- added a preset for xenoborgs (along side a secret version)
- the xenoborg rule will pull players that are ready and have toggled xenoborg in the antag selection in the character menu
- it's also possible to add the xenoborg rule in the middle of the round and it will spawn ghostroles
- if the mothership is destroyed, all xenoborgs explode
- if all xenoborgs explode, a announcement will inform the crew that the xenoborg threat is over (however it won't call evac)
- roundend summary information regarding if the xenoborgs won or lost
- guidebook entry for xenoborgs

<!-- What did you change? -->

## Why / Balance
check the [design dod](https://github.com/space-wizards/docs/blob/master/src/en/space-station-14/round-flow/antagonists/Xenoborgs.md)

<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->

## Technical details
- Changed some yml, sprites, added some adudio. but it is mostly C#
- Changed the Antag Selection to allow xenoborgs to work
- Added whitelist to spawn points so the mothership core would only spawn in the middle of the core
- Added Xenoborg component, as well as the xenoborg system, that deals with the mothership core being destroyed
- Added Xenoborg rule component and system.

<!-- Summary of code changes for easier review. -->

## Media
<img width="376" height="144" alt="image" src="https://github.com/user-attachments/assets/2cf1aac5-9568-48d1-9df9-9d91f9edc2fc" />
<img width="572" height="534" alt="image" src="https://github.com/user-attachments/assets/e6ee1071-b07a-4d2a-af6c-b58acf98916c" />
<img width="480" height="520" alt="image" src="https://github.com/user-attachments/assets/bf8972cb-4da5-4f61-9abf-91e02617260a" />

<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->

**Changelog**
:cl: Samuka, DarkIcedCoffee, Whale
- add: Added a new antag: Xenoborgs

<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
<!--
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->

Coolsurf6 commented:
LETS GOOOOO!

I'm hyped!

Samuka-C commented on Resources/Textures/Mobs/Silicon/mothership_core.rsi/meta.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "version": 1,
```
ah... I thought it was the sprites. welcome, going back to 1 then

Alpha-Two commented:
I think the guidebook entry, for borging crew at least, could be much improved.
The heading itself should not be the list introduction (perhaps just "Borging crew"?), moved to the text body instead.
Additionally, it should either be a numbered list, or there should be commas after each entry (with an "and" before the last entry)

Samuka-C commented:
> I think the guidebook entry, for borging crew at least, could be much improved. The heading itself should not be the list introduction (perhaps just "Borging crew"?), moved to the text body instead. Additionally, it should either be a numbered list, or there should be commas after each entry (with an "and" before the last entry)

like so? (I updated the PR comment image)

southbridge-fur reviewed (approved):
Just one question, everything looks good to me.

Samuka-C commented on Resources/Prototypes/Entities/Structures/Machines/Computers/computers.yml:
```diff
@@ -1574,20 +1574,28 @@
     transmitFrequencyId: Xenoborg
   - type: Computer
     board: ComputerXenoborgsControlCircuitboard
-  - type: AccessReader # only used for dangerous things
```
the "control" computer can't actually destroy or disable the xenoborgs, just check how many there are, and their battery and integrity.

I should remove the Access Reader

Samuka-C replied:
ahh, I already removed the AccessReader

Samuka-C commented:
> > > I think the guidebook entry, for borging crew at least, could be much improved. The heading itself should not be the list introduction (perhaps just "Borging crew"?), moved to the text body instead. Additionally, it should either be a numbered list, or there should be commas after each entry (with an "and" before the last entry)
> > 
> > 
> > like so? (I updated the PR comment image)
> 
> There's still many issues, here's a possible alternate version (no formatting):
> 
> # Adding crew to your ranks
> As the mothership, you must grow your army of borgs. To do so, you must convert all the sentient beings you can find. This can be achieved in the following manner:
> 
> 1. Have your Xenoborgs bring dead crew to your shop, placing them in the body crusher;
> 2. Print a Xenoborg shell from gathered parts using your fabricator and unlock it;
> 3. Open the Xenoborg with a crowbar, and remove the empty MMI inside;
> 4. If the victim's brain is organic, place it in the MMI you just removed;
> 5. Place the MMI - or the positronic brain, if the victim was a silicon - back into the Xenoborg;
> 6. Optionally, rename the Xenoborg and upgrade their modules;
> 7. Close the Xenoborg with the crowbar, and re-engage the lock.
> 
> The new Xenoborg will be uncharged, yet functional. They will need to go to a recharging station to restore their full power and capabilities.

done, thanks for the suggestion. maybe try looking at the rest of the guidebook too? 'v'

Samuka-C commented on Content.Server/Antag/Components/AntagSelectionComponent.cs:
```diff
@@ -201,6 +201,12 @@ public partial struct AntagSelectionDefinition()
+    /// Prototype of the entity this antag selection should use
+    /// </summary>
+    [DataField]
+    public EntProtoId? EntityPrototype;
```
The spawner prototype is for spawning the role as a ghostrole if the rule is added midround

Nukeops rule actually have another component that defines the entity the players will be (is their own saved character)

there is a rule that defines a single entity. so I could use it, but then it would spawn all xenoborgs of the same proto, so ops, all cores or some other xenoborg type.

This is the solution i came up with. if you can find one better, I would love to know

slarticodefast replied:
I think `AntagSpawnerComponent` is already exists to do exactly what you are doing here.

Samuka-C replied:
Doesn't work, cause it only defines one prototype for the whole rule, i need 5 prototypes, and also that each prototype is associated with a different selection, since players that ready up for core get spawned as the mothership core and same for the players that readied as xenoborgs

Samuka-C commented on Content.Shared/Xenoborgs/Components/MothershipCoreComponent.cs:
```diff
@@ -0,0 +1,13 @@
+    ///     Path to antagonist alert sound.
+    /// </summary>
+    [DataField]
+    public SoundSpecifier GreetSoundNotification = new SoundPathSpecifier("/Audio/Ambience/Antag/xenoborg_start.ogg");
```
the mothership core works for more than just the sounds, it serves to explode all xenoborgs when all mothership cores are destroyed. its also used by the xenoborg rule to determine the end result of the round.

slarticodefast replied:
Ok, fine to keep then, but use the existing code for the greeting notification.

Samuka-C replied:
already done

Samuka-C commented:
> Also, for this to be merged, could you update the design doc so it includes all design decisions that were made during the development? Especially why they were chosen the way they are. For example the pinpointer, hand whitelists, and general interactions like the conversion mechanics. Maybe a few comments on the art design as well. Just so that others can follow all of your decisions.

Well, I guess it will have to be another PR, since is a different, repo, but sure

Samuka-C commented on Resources/Prototypes/tags.yml:
```diff
@@ -949,6 +949,12 @@
+  id: MothershipModule
+
 - type: Tag
   id: FakeMindShieldImplant
```
done

Samuka-C commented on Content.Shared/Xenoborgs/Components/MothershipCoreComponent.cs:
```diff
@@ -0,0 +1,13 @@
+namespace Content.Shared.Xenoborgs.Components;
+
+[RegisterComponent]
+public sealed partial class MothershipCoreComponent : Component
```
is this enough?

ScarKy0 commented on Content.Shared/Silicons/Borgs/Components/BorgChassisComponent.cs:
```diff
@@ -78,6 +78,12 @@ public sealed partial class BorgChassisComponent : Component
     public ProtoId<AlertPrototype> NoBatteryAlert = "BorgBatteryNone";
+
+    /// <summary>
+    /// If the entity can open own UI (default: false)
```
```suggestion
    /// If the entity can open own UI.
```

ScarKy0 commented on Content.Shared/Xenoborgs/Components/MothershipCoreComponent.cs:
```diff
@@ -0,0 +1,16 @@
+[RegisterComponent]
+public sealed partial class MothershipCoreComponent : Component
+{
+    /// <summary>
+    ///     Path to antagonist alert sound.
+    /// </summary>
+    [DataField]
+    public SoundSpecifier GreetSoundNotification = new SoundPathSpecifier("/Audio/Ambience/Antag/xenoborg_start.ogg");
```
You can keep the component however this you can just replace by briefing in the rule and antag stuff. Look at how paradox clones play their briefing.

ScarKy0 commented on Content.Shared/Xenoborgs/Components/XenoborgComponent.cs:
```diff
@@ -1,7 +1,16 @@
-public sealed partial class XenoborgMothershipComponent : Component;
+public sealed partial class XenoborgComponent : Component
+{
+    /// <summary>
+    ///     Path to antagonist alert sound.
+    /// </summary>
+    [DataField]
+    public SoundSpecifier GreetSoundNotification = new SoundPathSpecifier("/Audio/Ambience/Antag/xenoborg_start.ogg");
```
Same

Samuka-C commented on Content.Shared/Xenoborgs/Components/XenoborgComponent.cs:
```diff
@@ -1,7 +1,13 @@
+    ///     Path to antagonist alert sound.
+    /// </summary>
+    [DataField]
+    public SoundSpecifier GreetSoundNotification = new SoundPathSpecifier("/Audio/Ambience/Antag/xenoborg_start.ogg");
```
same stuff, so... already done

Samuka-C commented:
Last commit should fix the issue [40105](https://github.com/space-wizards/space-station-14/issues/40105)

Samuka-C commented:
> Also, for this to be merged, could you update the design doc so it includes all design decisions that were made during the development? Especially why they were chosen the way they are. For example the pinpointer, hand whitelists, and general interactions like the conversion mechanics. Maybe a few comments on the art design as well. Just so that others can follow all of your decisions.

done

Samuka-C commented on Content.Server/Antag/Components/AntagSelectionComponent.cs:
```diff
@@ -201,6 +201,12 @@ public partial struct AntagSelectionDefinition()
+    /// Prototype of the entity this antag selection should use
+    /// </summary>
+    [DataField]
+    public EntProtoId? EntityPrototype;
```
> You could make multiple separate spawn points for this that each ensure the correct prototype being spawned.

Can you explain this solution a little more?

Samuka-C commented on Resources/ServerInfo/Guidebook/Antagonist/Xenoborgs.xml:
```diff
@@ -0,0 +1,167 @@
+  There is a total of four types of Xenoborgs. Each Borg type has certain abilities that the others lack and require teamwork to help grow the xenoborgs numbers.
+  Each type of Xenoborg has unique modules that only fit in that specific type. The Mothership Core can produce upgrade modules designed for specific Xenoborg types.
+
+  All Xenoborgs have a pin pointer pointing to the Mothership's location, a GPS, and a material bag for collecting materials for the creation of more xenoborgs, and basic tools.
```
done

ToastEnjoyer commented on Content.Server/Xenoborgs/XenoborgSystem.cs:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+
+    public void EnsureXenoborgRole(EntityUid mindId, EntityUid ent, XenoborgComponent xenoborgComp)
+    {
+        if (!_roles.MindHasRole<XenoborgRoleComponent>(mindId))
+        {
+            _roles.MindAddRole(mindId, "MindRoleXenoborg", silent: true);
+
+            if (TryComp<ActorComponent>(ent, out var actorComp))
+            {
+                _antag.SendBriefing(actorComp.PlayerSession,
+                    Loc.GetString("xenoborgs-welcome"),
+                    Color.BlueViolet,
+                    new SoundPathSpecifier("/Audio/Ambience/Antag/xenoborg_start.ogg")
+                );
+            }
+        }
+    }
```
Possibly add a comment to this

Samuka-C replied:
done

ToastEnjoyer commented on Resources/Prototypes/Entities/Markers/Spawners/Conditional/xenoborgs.yml:
```diff
@@ -0,0 +1,33 @@
+    layers:
+    - state: green
+    - sprite: Mobs/Silicon/chassis.rsi
+      state: xenoborg_engi
```
Is this the best state we could set it to? Mappers might get confused but it's not a big deal

ToastEnjoyer replied:
Probably good idea, isn't super necessary but up to you.

ToastEnjoyer commented on Resources/Prototypes/GameRules/subgamemodes.yml:
```diff
@@ -70,3 +70,45 @@
+      mindRoles:
+      - MindRoleXenoborg
+      min: 4
+      max: 4
```
Maybe we should increase the max

Samuka-C replied:
what do you suggest for the max? at 80 pop (80 ready people)

ToastEnjoyer replied:
I'd say 6-8 would be good

ToastEnjoyer commented on Resources/Prototypes/game_presets.yml:
Probably all at once because they are a major round antag

ToastEnjoyer reviewed (changes requested):
Took an extensive review of Xenoborgs, very well made. Only a few minor things in the requests

ToastEnjoyer commented on Resources/Prototypes/Roles/Antags/xenoborgs.yml:
```diff
@@ -0,0 +1,23 @@
+  objective: roles-antag-xenoborg-objective
+  requirements:
+  - !type:RoleTimeRequirement
+    role: JobBorg
```
> Well, AI can't control borgs so is not like a AI time requirement will make sure the player knows how to play this antag
> 
> 
> 
> maybe a sci time requirement , so they know how to borg people (even thou is not guaranteed people that play sci knows how to borg people)

I think there should be atleast some AI time, the experience is very different to Borg gameplay

Samuka-C commented on Content.Server/Antag/Components/AntagSpawnerComponent.cs:
```diff
@@ -1,4 +1,5 @@
 using Content.Server.Antag;
```
done

Samuka-C commented on Content.Server/Antag/AntagSelectionSystem.cs:
```diff
@@ -534,21 +535,21 @@ public bool IsSessionValid(Entity<AntagSelectionComponent> ent, ICommonSession?
-                    return false;
-                break;
-            }
+                {
```
well my IDE (vscode with C# extension) was complaining about it saying: [IDE0055](https://learn.microsoft.com/pt-br/dotnet/fundamentals/code-analysis/style-rules/ide0055)

Samuka-C commented on Content.Server/GameTicking/Rules/Components/XenoborgsRuleComponent.cs:
```diff
@@ -0,0 +1,4 @@
+namespace Content.Server.GameTicking.Rules.Components;
+
+[RegisterComponent, Access(typeof(NukeopsRuleSystem))]
```
done

Samuka-C commented on Content.Server/GameTicking/Rules/XenoborgsRuleSystem.cs:
```diff
@@ -0,0 +1,168 @@
+    /// Gets the number of humans who are alive
+    /// </summary>
+    /// <returns></returns>
+    private int GetNumberHumans()
```
done

Samuka-C commented on Content.Server/GameTicking/Rules/XenoborgsRuleSystem.cs:
```diff
@@ -0,0 +1,168 @@
+    /// </summary>
+    /// <param name="playerControlled">if it should only include xenoborgs with a mind</param>
+    /// <param name="alive">if it should only include xenoborgs that are alive</param>
+    /// <returns></returns>
```
done

Samuka-C commented on Content.Server/GameTicking/Rules/XenoborgsRuleSystem.cs:
```diff
@@ -0,0 +1,168 @@
+    /// <summary>
+    /// Gets the number of xenoborg cores
+    /// </summary>
+    /// <returns></returns>
```
done

Samuka-C commented on Content.Server/Antag/Components/AntagSpawnerComponent.cs:
```diff
@@ -14,4 +15,10 @@ public sealed partial class AntagSpawnerComponent : Component
+    /// List of different entity prototypes for each antag role that is selected by the <see cref="AntagSelectionSystem"/>.
+    /// </summary>
+    [DataField]
+    public Dictionary<ProtoId<AntagPrototype>, List<EntProtoId>>? Prototypes;
```
Okay, you fixed this

Samuka-C commented on Content.Server/GameTicking/Rules/XenoborgsRuleSystem.cs:
```diff
@@ -0,0 +1,168 @@
+    {
+        _chatSystem.DispatchGlobalAnnouncement(
+            Loc.GetString("mothership-destroyed-announcement"),
+            colorOverride: Color.Gold);
```
done

Samuka-C commented on Content.Server/GameTicking/Rules/XenoborgsRuleSystem.cs:
```diff
@@ -0,0 +1,168 @@
+    [Dependency] private readonly AntagSelectionSystem _antag = default!;
+    [Dependency] private readonly ChatSystem _chatSystem = default!;
+    [Dependency] private readonly MobStateSystem _mobState = default!;
+    [Dependency] private readonly SharedMindSystem _mindSystem = default!;
```
done

Samuka-C commented on Content.Server/Antag/AntagSelectionSystem.cs:
```diff
@@ -603,10 +604,12 @@ private void OnObjectivesTextGetInfo(Entity<AntagSelectionComponent> ent, ref Ob
 {
     public readonly ICommonSession? Session = Session;
 
+    public readonly List<ProtoId<AntagPrototype>>? AntagRoles = AntagRoles;
```
you fixed it

Samuka-C commented on Content.Server/Xenoborgs/XenoborgSystem.cs:
```diff
@@ -0,0 +1,104 @@
+            {
+                _antag.SendBriefing(actorComp.PlayerSession,
+                    Loc.GetString("mothership-welcome"),
+                    Color.BlueViolet,
```
done

Samuka-C commented on Content.Server/GameTicking/Rules/XenoborgsRuleSystem.cs:
```diff
@@ -0,0 +1,168 @@
+
+        _chatSystem.DispatchGlobalAnnouncement(
+            Loc.GetString($"xenoborgs-no-more-threat-mothership-core-{status}-announcement"),
+            colorOverride: Color.Gold);
```
done

Samuka-C commented on Content.Server/Xenoborgs/XenoborgSystem.cs:
```diff
@@ -0,0 +1,104 @@
+            {
+                _antag.SendBriefing(actorComp.PlayerSession,
+                    Loc.GetString("xenoborgs-welcome"),
+                    Color.BlueViolet,
```
done

Pok27 commented:
@Samuka-C Please do not write in every review that it has been resolved, spamming the DS channel. Just click “Resolve conversation,” and the reviewer will check it.

Samuka-C commented:
> @Samuka-C Please do not write in every review that it has been resolved, spamming the DS channel. Just click “Resolve conversation,” and the reviewer will check it.

ah sorry

beck-thompson commented on Content.Server/Antag/Components/AntagSelectionComponent.cs:
```diff
@@ -47,7 +47,7 @@ public sealed partial class AntagSelectionComponent : Component
     /// </summary>
     [DataField]
-    public Dictionary<AntagSelectionDefinition, HashSet<ICommonSession>>PreSelectedSessions = new();
+    public Dictionary<AntagSelectionDefinition, HashSet<ICommonSession>> PreSelectedSessions = new();
```
undo this (I'm only saying this because the file is otherwise untouched)

beck-thompson commented on Content.Server/Antag/Components/AntagSpawnerComponent.cs:
```diff
@@ -1,4 +1,4 @@
-using Content.Server.Antag;
+using Content.Shared.Roles;
```
same here (untouch)

Samuka-C replied:
Shared.Roles is necessary? the code shows errors if i remove it

beck-thompson commented on Content.Server/Antag/AntagSpawnerSystem.cs:
```diff
@@ -1,3 +1,4 @@
+using System.Linq;
```
same

Samuka-C replied:
ah, I was checking the wrong file

beck-thompson commented on Content.Server/Silicons/Borgs/BorgSystem.Transponder.cs:
```diff
@@ -118,8 +118,11 @@ private void Disable(Entity<BorgTransponderComponent, BorgChassisComponent?> ent
     }
 
-    private void Destroy(Entity<BorgTransponderComponent> ent)
+    public void Destroy(Entity<BorgTransponderComponent?> ent)
```
add a comment now this is public

beck-thompson commented on Content.Server/Silicons/Laws/SiliconLawSystem.cs:
```diff
@@ -2,6 +2,7 @@
 using Content.Server.Administration;
 using Content.Server.Chat.Managers;
 using Content.Server.Station.Systems;
+using Content.Server.Xenoborgs;
```
remove all changes from this file as well

beck-thompson commented on Content.Server/Xenoborgs/XenoborgSystem.cs:
```diff
@@ -0,0 +1,110 @@
+        if (_roles.MindHasRole<XenoborgRoleComponent>(args.Mind))
+            return;
+
+        _roles.MindAddRole(args.Mind, "MindRoleXenoborg", silent: true);
```
probably best to put the string in the component

beck-thompson commented on Content.Server/Xenoborgs/XenoborgSystem.cs:
```diff
@@ -0,0 +1,110 @@
+        if (_roles.MindHasRole<XenoborgCoreRoleComponent>(args.Mind))
+            return;
+
+        _roles.MindAddRole(args.Mind, "MindRoleMothershipCore", silent: true);
```
same

beck-thompson commented on Content.Server/Xenoborgs/XenoborgSystem.cs:
```diff
@@ -0,0 +1,110 @@
+        _antag.SendBriefing(actorComp.PlayerSession,
+            Loc.GetString("xenoborgs-welcome"),
+            XENOBORG_BRIEFING_COLOR,
+            new SoundPathSpecifier("/Audio/Ambience/Antag/xenoborg_start.ogg")
```
put the sound specifier in the comp probably

beck-thompson commented on Content.Server/GameTicking/Rules/XenoborgsRuleSystem.cs:
```diff
@@ -0,0 +1,144 @@
+
+        return numberMothershipCores;
+    }
+
```
remove space

beck-thompson reviewed (changes requested):
a few more things I noticed

beck-thompson commented on Content.Server/Antag/Components/AntagSpawnerComponent.cs:
```diff
@@ -1,4 +1,3 @@
-using Content.Server.Antag;
```
Remove (only because you didn't touch the file besides this)

beck-thompson commented on Content.Server/Antag/Components/AntagMultipleRoleSpawnerComponent.cs:
```diff
@@ -0,0 +1,23 @@
+public sealed partial class AntagMultipleRoleSpawnerComponent : Component
+{
+    /// <summary>
+    ///     Antag prototype -> List of entities to spawn for that prototype
```
do something like this:

antag prototype -> list of possible entities to spawn for that antag prototype. Will choose from the list randomly once with replacement unless <see cref="PickAndTake"> is set to true

beck-thompson commented on Content.Server/Antag/Components/AntagMultipleRoleSpawnerComponent.cs:
```diff
@@ -0,0 +1,23 @@
+namespace Content.Server.Antag.Components;
+
+/// <summary>
+/// Spawns multiple prototypes for each antag prototype selected by the <see cref="AntagSelectionSystem"/>
```
change to something like:

selects one prototype from a list for each antag prototype

it sounds like it spawns multiple entities right now

beck-thompson commented on Content.Server/Antag/AntagSelectionSystem.cs:
```diff
@@ -616,12 +619,14 @@ public record struct AntagSelectEntityEvent(ICommonSession? Session, Entity<Anta
 
     public bool Handled => Coordinates.Any();
 
+    public EntityUid Entity = Entity;
```
add comment

beck-thompson commented on Content.Server/Antag/AntagSelectionSystem.cs:
```diff
@@ -603,10 +604,12 @@ private void OnObjectivesTextGetInfo(Entity<AntagSelectionComponent> ent, ref Ob
 {
     public readonly ICommonSession? Session = Session;
 
+    public readonly List<ProtoId<AntagPrototype>> AntagRoles = AntagRoles;
```
add comment

beck-thompson commented on Content.Server/GameTicking/Rules/XenoborgsRuleSystem.cs:
```diff
@@ -0,0 +1,149 @@
+        var xenoborgQuery = AllEntityQuery<XenoborgComponent>();
+        while (xenoborgQuery.MoveNext(out var xenoborgEnt, out _))
+        {
+            if (TryComp<MothershipCoreComponent>(xenoborgEnt, out _))
```
use hascomp (we talked about this in discord)

beck-thompson commented on Content.Server/GameTicking/Rules/XenoborgsRuleSystem.cs:
```diff
@@ -0,0 +1,149 @@
+        var query = AllEntityQuery<XenoborgComponent>();
+        while (query.MoveNext(out var xenoborg, out _))
+        {
+            if (TryComp<MothershipCoreComponent>(xenoborg, out _))
```
same - you might also want to have a new component just for xenoborgs. E.g all xenoborgs would have the xenoborg component and then the queen has a unique one and the babies have a unique one as well. That might be a little nicer 🤔

I think though this is probably fine though because I think the queen should only ever be the one excluded

beck-thompson commented on Content.Server/Silicons/Laws/SiliconLawSystem.cs:
```diff
@@ -88,7 +88,6 @@ private void OnLawProviderMindRemoved(Entity<SiliconLawProviderComponent> ent, r
         if (!ent.Comp.Subverted)
             return;
         RemoveSubvertedSiliconRole(args.Mind);
-
```
put back

beck-thompson commented on Content.Server/Spawners/Components/SpawnPointComponent.cs:
```diff
@@ -19,6 +20,12 @@ public sealed partial class SpawnPointComponent : Component, ISpawnPoint
+    /// Whitelist of entities that can be spawned at this SpawnPoint
+    /// </summary>
+    [DataField]
+    public EntityWhitelist? SpawnWhitelist;
```
I think it would be better to move this into its own component instead of adding it here. then in the rules grid system just add a trycomp (we are lazy so its probably fine) and check if there is a whitelist. Adding it to the spawn point component itself is sussy because the whitelist is not used anywhere besides the grid system (`SpawnPointSystem` doesn't use it)

<img width="979" height="603" alt="Image" src="https://github.com/user-attachments/assets/d2a77323-3456-475c-b9b6-23854a945abe" />

beck-thompson commented on Content.Shared/Xenoborgs/Components/XenoborgComponent.cs:
```diff
@@ -1,7 +1,32 @@
+    /// The text that is sent when you become a xenoborg
+    /// </summary>
+    [DataField]
+    public string BriefingText = "xenoborgs-welcome";
```
Use `LocId` instead of `string`

beck-thompson commented on Resources/Prototypes/tags.yml:
```diff
@@ -1541,6 +1547,9 @@
 - type: Tag
   id: WriteIgnoreStamps
 
+- type: Tag
```
I think you can use the components instead of the tags for the whitelist spawns

beck-thompson commented on Content.Server/Xenoborgs/XenoborgSystem.cs:
```diff
@@ -0,0 +1,81 @@
+        var xenoborgQuery = AllEntityQuery<XenoborgComponent, BorgTransponderComponent>();
+        while (xenoborgQuery.MoveNext(out var xenoborgEnt, out _, out _))
+        {
+            if (TryComp<MothershipCoreComponent>(xenoborgEnt, out _))
```
Use Hascomp

beck-thompson commented on Content.Server/Spawners/Components/SpawnPointComponent.cs:
```diff
@@ -1,4 +1,5 @@
 using Content.Shared.Roles;
+using Content.Shared.Whitelist;
```
remove

beck-thompson replied:
remove

beck-thompson commented on Content.Server/GameTicking/Rules/RuleGridsSystem.cs:
```diff
@@ -54,7 +54,7 @@ private void OnLoadedGrids(Entity<RuleGridsComponent> ent, ref RuleLoadedGridsEv
     {
         var query = EntityQueryEnumerator<SpawnPointComponent, TransformComponent>();
-        while (query.MoveNext(out var uid, out _, out var xform))
+        while (query.MoveNext(out var uid, out var spawnPointComp, out var xform))
```
You don't need the `spawnPointComp` anymore

beck-thompson commented on Content.Server/GameTicking/Rules/RuleGridsSystem.cs:
```diff
@@ -65,6 +65,15 @@ private void OnSelectLocation(Entity<RuleGridsComponent> ent, ref AntagSelectLoc
             if (_whitelist.IsWhitelistFail(ent.Comp.SpawnerWhitelist, uid))
                 continue;
 
+            if (TryComp<GridSpawnPointWhitelistComponent>(ent, out var gridSpawnPointWhitelistComponent))
```
sus but I think we just gotta take it at this point

Princess-Cheeseballs commented:
> I also couldn't control the ship manually, only FTL. Is that intended?

Might've been caused by a prototype refactor Slarti and I did? But I haven't heard any issues about it reported. If so it just needs to inherit from BaseMob again and then have BlockMovement added as a component with bypass bools.

Samuka-C commented on Content.Server/GameTicking/Rules/XenoborgsRuleSystem.cs:
```diff
@@ -102,6 +105,30 @@ protected override void AppendRoundEndText(EntityUid uid,
+    protected override void ActiveTick(EntityUid uid, XenoborgsRuleComponent component, GameRuleComponent gameRule, float frameTime)
+    {
+        base.ActiveTick(uid, component, gameRule, frameTime);
+        if (!component.NextRoundEndCheck.HasValue || component.NextRoundEndCheck > _timing.CurTime)
```
what do you mean?

Samuka-C commented:
> Also update the guidebook to have the new victory condition you added (looks like its more than 70% of crew is xenoborgs)

This is not a win condition, is a condition to call evac, the win condition is to have more than 50% of the crew turned into xenoborgs, which is already told in the guidebook

```
[bold][color=#2288ff]Xenoborg Major Victory![/color][/bold]
- There are more xenoborgs than alive crew in the end.
```

Road-Train commented:
> That invincibility bug still seems to an issue, you can't shoot the core directly with guns. You can damage it with melee weapons, but because of its damage reductions it takes quite a while.
> 
> I also couldn't control the ship manually, only FTL. Is that intended?
> 
> I'm was also having some trouble making borgs: <img alt="image" width="693" height="317" src="https://private-user-images.githubusercontent.com/107373427/494626172-37791c6c-d5db-44c4-8f6c-935fde5b812d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTkxNDQwMjQsIm5iZiI6MTc1OTE0MzcyNCwicGF0aCI6Ii8xMDczNzM0MjcvNDk0NjI2MTcyLTM3NzkxYzZjLWQ1ZGItNDRjNC04ZjZjLTkzNWZkZTViODEyZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkyOVQxMTAyMDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYTZhNGZhYmM1NTA0NzkwZDk1YWNlMGYzZTYyNzFiZTEyMGZhMThlYzQyYjI0NzBkNmVjNzgxOThhZWIzOWVmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.4ydpm4NoVI-UzgVwJHrP-3MeME7D400bnK6D4uqU1vE">
> 
> Where exactly is the fabricator? I can't seem to find it - maybe add a picture to the guidebook?
> 
> Also update the guidebook to have the new victory condition you added (looks like its more than 70% of crew is xenoborgs)

The mothership core is actually an exofab, it can make borgs.

beck-thompson commented on Content.Server/GameTicking/Rules/Components/XenoborgsRuleComponent.cs:
```diff
@@ -0,0 +1,29 @@
+    /// The amount of time between each check for the end of the round.
+    /// </summary>
+    [DataField]
+    public TimeSpan EndCheckDelay = TimeSpan.FromSeconds(30);
```
maybe decrease this to 15 seconds

beck-thompson commented:
Lets get this bad boy in game
